### DISCUSSION TITLE
Conditional Rendering

### DISCUSSION THREAD

**Discussion by m4rrc0** [withastro/roadmap#159] [Proposal]
In a component with facultative props, we might have to do a lot of data checking before rendering. This is really cumbersome.
I am wondering is this is something Astro could/should simplify.

Here is an example of a component I have been creating lately:
```
---
// Ellipsis
---
{hasSomethingToShow && (
    <div>
        {title && (
            <h2>{title}</h2>
        )}
        {subtitle && (
            <p>{subtitle}</p>
        )}
        {(featured_image?.src || text) && (
            <div>
                {featured_image?.src && (
                    <img src={featured_image?.src} />
                )}
                {text && (
                    <Markdown content={text} />
                )}
            </div>
        )}
    </div>
)}
```

The way I see it, Astro could detect a specific prop or directive. If what we pass to it is `null` or `undefined`, it does not render the element or component.
```
<img skip:val={featured_image?.src} src={featured_image?.src} />
```

Optionally, it could be extra awesome to have a way to trigger this behavior if the inner content passed is `null` or `undefined`.
```
<p skip:inner>{subtitle}</p>
```

Then, the above component would become
```
<div skip:inner>
    <h2 skip:inner>{title}</h2>
    <p skip:inner>{subtitle}</p>
    <div skip:inner>
        <img skip:val={featured_image?.src} src={featured_image?.src} />
        <Markdown skip:val={text} content={text} />
    </div>
</div>
```

**Comment by danilopolani** [comment]
This make sense, but at the same time you could just use one of the 6 supported frameworks integrations to do this, since Astro it's not much about "templating" (?)

**Comment by natemoo-re** [comment]
This seems like a cool idea! Similar to Vue's `v-if`. I definitely agree that the JSX approach of `&&` everywhere isn't the most ergonomic.

Not sure I love the `skip:` syntax proposed here, but maybe something like [Solid's `Show` component](https://www.solidjs.com/tutorial/flow_show)?

**Reply by okikio** [reply]
I like the Show syntax

**Reply by natemoo-re** [reply]
Thinking about it more, the `skip:inner` idea is super nice and I don't think I've ever seen a magical `if` like that anywhere else... Hmm...

**Reply by m4rrc0** [reply]
I am not sold on the `skip:` syntax either. It was just what seemed more appropriate when I thought about it. I first tried to find a 'positive' word but couldn't find anything as clear. For example, `show` does not seem as clear to me. 'Show' is the opposite of 'hide', so is it hidden like in `visibility: hidden;` or `display: none;`?
Anyway, I am curious to hear more thoughts on the syntax as well.

**Reply by m4rrc0** [reply]
The `skip:inner` idea seems indeed super powerful. It is so relaxing to know that the parent component will show if any child is shown. You can mess with the children in any way you like without ever having to change the condition on the parent.

And I have to say I am pleased with how easy it was to check for content (with my `io` function). I may be overlooking some cases (like named slots) but it does not seem that bad to handle.

**Reply by m4rrc0** [reply]
To comment further on the syntax, I think the most explicit phrasing would be something like "render if".
If implemented as a directive, maybe something like `render:if` would skip the element it there is no inner content and `render:if={someValue}` to check a specific value.

Maybe the `render` directive can be used for other functionality (with another keyword than `if`) later as well... (Maybe a `render:await` someday for example)

One other question comes to mind: should we provide a `render:else` like what Vue does?
It might be useful, especially with the inner content check idea. It would defeat the whole purpose of this idea if we end up having to 'reverse check' all the inner values of the preceding element to show something else in place.

**Comment by m4rrc0** [comment]
I am glad this idea triggers some interest. :)
I've been playing with it the other day and implemented a function to be plugged in any component.
```
async function io(Astro) {
  const { slots } = Astro
  const { $skip, $skipNoInner, ...props } = Astro.props

  let content = ''
  if (slots.has('default')) {
    content = (await slots?.render('default')) || ''
  }

  const $render =
    !($skip === true) && !($skipNoInner === true && content === '')

  return { ...Astro, props, $render }
}
```
Note that this is a simple POC, it needs improvements, for example it needs a better handling of `slots` if it is to be used with named slots.

We can use it on any component to conditionally render the whole component.
For example, we can create a `<Div />` component like so.
```
---
import { io } from '@utils'
const { props, $render } = await io(Astro)
---
{$render && <div {...props}><slot /></div>}
```
I wanted to play with the idea a little bit more before replacing every possible HTML element with my own component so I created just one component to catch any html element.
```
---
import { io } from '@utils'
const { props: allProps, $render } = await io(Astro)

const { div, p, span, img, h1, h2, /* Ellipsis */ ...props } = allProps
---

{div && $render && <div {...props}><slot /></div>}
{p && $render && <p {...props}><slot /></p>}
{span && $render && <span {...props}><slot /></span>}
{img && $render && <img {...props}><slot /></img>}
{h1 && $render && <h1 {...props}><slot /></h1>}
{h2 && $render && <h2 {...props}><slot /></h2>}
<!-- Ellipsis -->
```

I wanted to see as clearly and as early as possible the element name when using this component, that is why I didn't go for a `tag` prop for example.

Now the example I provided in my original post has become:
```
---
import { IO, Md } from '@components'
// Ellipsis
---
<IO div $skipNoInner>
    <IO h2 $skipNoInner>{title}</IO>
    <IO p $skipNoInner>{subtitle}</IO>
    <IO div $skipNoInner>
        <IO img $skip={!featured_image?.src} src={featured_image?.src} class="" style="max-width: 100px;" />
        <Md $skip={!text} content={text} />
    </IO>
</IO>
```
Note that I wrapped the astro `<Markdown />` component within my `<Md />` component using the `io` function.

**Reply by natemoo-re** [reply]
Here's a translation of your idea with a new primitive called `maybe` that I've been exploring.

https://stackblitz.com/edit/astro-maybe?file=src%2Fpages%2Findex.astro

**Reply by natemoo-re** [reply]
I think I settled on a syntax I like! This is a `maybe` component that renders under one of two conditions:

1. By default, when any children exist.
2. Optionally, when a specific `@when` condition is met. This could definitely be any other type of attribute.

```astro
<maybe.div class="wrapper">
    <maybe.h2>{title}</maybe.h2>
    <maybe.p>{subtitle}</use.p>
    <maybe.div>
      <maybe.img @when={featured_image?.src} src={featured_image.src} style="max-width: 500px;" />
    </maybe.div>
</maybe.div>
```

This is very in the weeds, but here's the Astro component implementation.

```astro
---
const { '@when': value, as: Child = Fragment, ...props } = Astro.props;
const hasWhen = '@when' in Astro.props;

let html: string;
if (hasWhen && value && Astro.slots.has('default')) {
  html = await Astro.slots.render('default');
  if (!html.trim()) html = undefined;
}
if (!hasWhen && Astro.slots.has('default')) {
  html = await Astro.slots.render('default');
  if (!html.trim()) html = undefined;
}
---

{hasWhen && value && <Child {...props} set:html={html} />}
{!hasWhen && html && <Child {...props} set:html={html} />}
```

...and the JavaScript wrapper that generates a `div` from `maybe.div`.

```js
import Maybe from './Maybe.astro';
const create = (name) => {
  const component = (res, props, ...args) =>
    Maybe(res, { as: name, ...props }, ...args);
  // this is technically private, but it signals to Astro that this JS function is an Astro component!
  component.isAstroComponentFactory = true;
  return component;
};
export const maybe = new Proxy(
  {},
  {
    get(_, name) {
      return create(name);
    },
  }
);
```

**Reply by okikio** [reply]
@natemoo-re I'd personally prefer if `@when` was just built-into all elements.
Plus what if you have a custom React element would `<maybe.ReactElement></maybe.ReactElement>` work

**Reply by m4rrc0** [reply]
> I'd personally prefer if @when was just built-into all elements.

I think we are exploring the idea with custom made components before considering if it should go in the Astro core. So far, I am personally 100% sold on the idea. It has made my components a lot more readable.

@natemoo-re Wow! Where can I learn more about the inner workings of Astro? I had no idea we could do that! Thanks.

**Reply by m4rrc0** [reply]
> 1. By default, when any children exist.

Oh and I definitely agree with this approach my `$skipNoInner` is a waste of room. ^^

**Reply by natemoo-re** [reply]
@m4rrc0 Yep, I think this is a great way to explore the ideas before bringing them up for an RFC! I should have some more time after launch week to create an `astro-flow` package with some helpers like this!

This is all pretty deep into the Astro internals, so we don't have very good docs on this kind of thing yet. But I'd love to share whatever knowledge I can and hopefully start documenting some of this stuff! ☺️
